File: src/main/java/com/dotgoing/utils/option/Some.java
package com.dotgoing.utils.option;

import java.util.function.Function;

public final class Some<T> extends Option<T> {
    private final T t;

    public Some(T t) {
        this.t = t;
    }

    @Override
    public <R> Option<R> map(Function<? super T, ? extends R> transformer) {
        return new Some<>(transformer.apply(value()));
    }

    @Override
    public <R> Option<R> flatMap(Function<? super T, ? extends Option<? extends R>> transformer) {
        return (Option<R>) transformer.apply(value());
    }

    @Override
    public T value() {
        return t;
    }

    @Override
    public boolean hasValue() {
        return true;
    }

    @Override
    public String toString() {
        return "Some(" + t + ")";
    }
}

File: src/main/java/com/dotgoing/utils/core/option/Some.kt
package com.dotgoing.utils.core.option

class Some<out T>(private val t: T) : Option<T>() {
    override fun hasValue(): Boolean {
        return true
    }

    override fun value(): T {
        return t
    }

    override fun errorCode(): Int {
        throw Exception("this data is ok: ${this}")
    }

    override fun error(): Exception {
        throw OptionStateException("Some obj has no error : ${this}")
    }

    override fun toString(): String {
        return try {
            t as String
        } catch (e: Exception) {
            "Some($t)"
        }

    }
}